### PR TITLE
Allow untyped imports

### LINKED ISSUE #11106: Better handling of untyped node module imports
Consider importing an untyped module, eg. exceljs:

```ts
import exceljs = require('exceljs');
```

This results in an error that the module exceljs is not found. There is at least two ways to fix this:

1. Setting `allowJs: true` and `maxNodeModulesJsDepth: 1`
2. Declare an ambient module in a separate file

I think it would be better if instead we update module resolution semantics such that when we find an appropriate package.json without a typings entry, we store it away. Then if module resolution fails (ie. we also can't find an @types/exceljs), we go back and import the untyped package.json as a JS import. Thoughts?

You might argue that this should only be allowed with --allowJs, but it might also make sense to just always use this behavior when using node resolution rules as you may still not want to allow JS in your own project while still allowing your dependencies to be authored in JS without typings.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #11106

If we find a `package.json` while searching through `node_modules`, we will allow imports of that module to work in an untyped fashion.
This change will not affect users with `--noImplicitAny` turned on.
It also has no effect without `--moduleResolution node`. So we might not want to completely close the issue until we have solutions for other module systems.

We can't do find-all-references for untyped imports because that function uses `symbol.declarations`, which in this case is empty.

### CODE REVIEW

**Comment by sheetalkamat** [comment]
Can you please add these test cases so we have baselines later if behavior modifies
- allowJS true and resolutuion happens to be untyped package
- allowjs true and resolution happens to be .js file
- allowjs false and there exists .js file file in the package.

**Comment by andy-ms** [comment]
| What | Where |
| --- | --- |
| allowJS true and resoltuion happens to be untyped package | untypedModuleImport_allowJs.ts |
| allowjs true and resolution happens to be .js file | moduleResolution_allowJs.ts |
| allowjs false and there exists .js file file in the package. | untypedModuleImport.ts |

**Comment by andy-ms** [comment]
I do find it suspicious that in `moduleResolution_allowJs.types`, we have `exports.default : any`, but when we import it we get `foo: { bar(): number }`. But that is a JS typings issue.

**Comment by DanielRosenwasser** [comment]
I don't think we should just be reporting `Cannot find module 'foo'.` to users - clearly we found something but the user is going to get undesirable behavior for their workflow. The correct thing to do is to explicitly tell the user that the module was found but that its is untyped/its shape has type `any`. I'd do something like

```
The module '{0}' resolved to a JavaScript package, but implicitly has type 'any'.
  Consider adding type declarations for '{0}' or using the '--allowJs' compiler option.
```

**Comment by andy-ms** [comment]
I went with:
```
A package for '{0}' was found at '{1}', but is untyped. Since '--noImplicitAny' is enabled, this package must have a declaration.
```

`--allowJs` would also require people to use `--nodeModulesMaxDepth 1`.

We could mention that option, but this error message will be seen very often by people who simply forget to install types. Maybe we could even recommend `npm install @types/foo` in the error message (though typings for it might not exist).

**Comment by DanielRosenwasser** [comment]
@andy-ms Nice - that one seems reasonable.

**Comment by SamVerschueren** [comment]
Nice work @andy-ms ! I wrote a blogpost about this a couple of weeks ago about [The one thing that keeps me from using TypeScript](https://medium.com/@sam.verschueren/the-one-thing-that-keeps-me-from-using-typescript-839e9bd464d7). This is exactly what I wanted! Also with the error message that @DanielRosenwasser suggested. The only difference is that I suggested an extra flag called `noImplicitAnyModule`. But using `noImplicitAny` might be even better!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1355,15 +1355,34 @@ namespace ts {
...
+                Debug.assert(!isRelative); // Only global modules may be untyped
+                Debug.assert(quotedName !== undefined);
+                if (compilerOptions.noImplicitAny) {
+                    error(errorNode,
+                        Diagnostics.A_package_for_0_was_found_at_1_but_is_untyped_Since_noImplicitAny_is_enabled_this_package_must_have_a_declaration,
+                        moduleReference,
+                        resolvedModule.resolvedFileName);
+                }
+
+                const newSymbol = createSymbol(SymbolFlags.ValueModule, quotedName);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: what happenes if you just return undefined? why do you need to create a symbol?

Comment (mhegazy) [inline src/compiler/checker.ts]: If you must, i would create a singleton `UnkownModuleSymbol` and check for that in `isUntypedModuleSymbol`

File: src/compiler/moduleNameResolver.ts [hunk 2]
```diff
@@ -650,57 +661,90 @@ namespace ts {
...
+    function loadModuleFromNodeModulesAtTypes(moduleName: string, directory: string, failedLookupLocations: string[], state: ModuleResolutionState): string | undefined {
+        return discardUntypedResults(loadModuleFromNodeModulesWorker(moduleName, directory, failedLookupLocations, state, /*checkOneLevel*/ false, /*typesOnly*/ true));
     }
 
-    function loadModuleFromNodeModulesWorker(moduleName: string, directory: string, failedLookupLocations: string[], state: ModuleResolutionState, checkOneLevel: boolean, typesOnly: boolean): string {
+    function loadModuleFromNodeModulesWorker(
+        moduleName: string, directory: string, failedLookupLocations: string[], state: ModuleResolutionState, checkOneLevel: boolean, typesOnly: boolean): ResolutionResult {
         directory = normalizeSlashes(directory);
+        /**
+         * Remembers the path to a package.json for the package.
```
Comment (DanielRosenwasser) [inline src/compiler/moduleNameResolver.ts]: "The path to a package.json in case we find a typed version of the package further up the spine."

**Comment by andy-ms** [comment]
Closing in favor of #11889

**Pull request closed without merging**